### PR TITLE
Fix default search field on dynamic mapping

### PR DESCRIPTION
**Pull request by evanxg852000**
Checks if default search is not specified and query has no search field declared for dynamic mapping.

Closes https://github.com/quickwit-oss/quickwit/issues/1526

### CODE REVIEW

File: quickwit-doc-mapper/src/query_builder.rs [hunk 1]
```diff
@@ -65,8 +74,27 @@ fn has_range_clause(user_input_ast: UserInputAst) -> bool {
...
             false
         }
-        UserInputAst::Boost(ast, _) => has_range_clause(*ast),
-        UserInputAst::Leaf(leaf) => matches!(*leaf, UserInputLeaf::Range { .. }),
+        UserInputAst::Boost(ast, _) => has_range_clause(ast),
+        UserInputAst::Leaf(leaf) => matches!(**leaf, UserInputLeaf::Range { .. }),
+    }
+}
+
+fn has_search_field(user_input_ast: &UserInputAst) -> bool {
```
Comment (fulmicoton) [inline quickwit-doc-mapper/src/query_builder.rs:82]: We should reject queries with the same error message if a query contains a single literal with no field.

In this PR, the query `hello` is rejected.
But the query `hello AND my_field:toto` is not rejected.

We want to reject both with this error message.

To fix it, you just need to change the code for `Clause` below, and make the logic stronger.  

```
 for (_, sub_ast) in sub_queries {
                if !has_search_field(sub_ast) {
                    return false;
                }
            }
            true
```

Comment (fulmicoton) [reply quickwit-doc-mapper/src/query_builder.rs:82]: And probably rename the function `has_search_field_everywhere` or a better name you can think of

Comment (fulmicoton) [reply quickwit-doc-mapper/src/query_builder.rs:82]: a unit test for this case would be good to :)

**Review by fulmicoton — CHANGES_REQUESTED** [review]
added comment/

File: quickwit-doc-mapper/src/query_builder.rs (hunk 1 above)
Comment (evanxg852000) [reply quickwit-doc-mapper/src/query_builder.rs]: No sure what you meant by "make the logic stronger" but I refactored and renamed from `has_search_field` to `needs_default_search_field`

Comment (fulmicoton) [reply quickwit-doc-mapper/src/query_builder.rs]: Here by "stronger", I meant it as make "the predicate stricter".

File: quickwit-doc-mapper/src/query_builder.rs [hunk 2]
```diff
@@ -65,8 +74,29 @@ fn has_range_clause(user_input_ast: UserInputAst) -> bool {
...
-        UserInputAst::Boost(ast, _) => has_range_clause(*ast),
-        UserInputAst::Leaf(leaf) => matches!(*leaf, UserInputLeaf::Range { .. }),
+        UserInputAst::Boost(ast, _) => has_range_clause(ast),
+        UserInputAst::Leaf(leaf) => matches!(**leaf, UserInputLeaf::Range { .. }),
+    }
+}
+
+/// Tells if the query has a Term or Range node which does not
+/// specify a search field.
+fn needs_default_search_field(user_input_ast: &UserInputAst) -> bool {
```
Comment (fulmicoton) [inline quickwit-doc-mapper/src/query_builder.rs:84]: good name

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**